jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java

```diff
@@ -109,6 +109,21 @@ private static native void checkpointJVMImpl(String imageDir,
 		});
 	}
 
+	/**
+	 * A hook can be one of the following states:
+	 * SINGLE_THREAD - a hook running when the single-threaded mode is enabled
+	 * MULTI_THREAD - a hook ruuning when the single-threaded mode is NOT enabled
+	 */
+	public static enum HookMode {
+		SINGLE_THREAD,
+		MULTI_THREAD
+	}
+
+	// the highest priority of a MULTI_THREAD hook
```

Comment:
we should put this with the rest of the priority constants

Reply:
These should be set relative to `MULTI_THREADED_HOOK_PRIORITY_HIGHEST`

```
	private static final int RESTORE_CLEAR_INETADDRESS_CACHE_PRIORITY = 100;
	private static final int RESTORE_ENVIRONMENT_VARIABLES_PRIORITY = 100;
	private static final int RESTORE_SYSTEM_PROPERTIES_PRIORITY = 100;
```

Reply:
Updated current priorities:
```
	// the lowest priority of a CONCURRENT_MODE hook
	public static int CONCURRENT_MODE_HOOK_PRIORITY_LOWEST = 0;
	// the highest priority of a CONCURRENT_MODE hook
	public static int CONCURRENT_MODE_HOOK_PRIORITY_HIGHEST = 99;
	// the lowest priority of a SINGLE_THREAD_MODE hook
	private static final int USER_HOOKS_PRIORITY = CONCURRENT_MODE_HOOK_PRIORITY_HIGHEST + 1;
	// other SINGLE_THREAD_MODE hooks
	private static final int RESTORE_CLEAR_INETADDRESS_CACHE_PRIORITY = CONCURRENT_MODE_HOOK_PRIORITY_HIGHEST + 99;
	private static final int RESTORE_ENVIRONMENT_VARIABLES_PRIORITY = CONCURRENT_MODE_HOOK_PRIORITY_HIGHEST + 99;
	private static final int RESTORE_SYSTEM_PROPERTIES_PRIORITY = CONCURRENT_MODE_HOOK_PRIORITY_HIGHEST + 99;
	/* RESET_CRIUSEC_PRIORITY and RESTORE_SECURITY_PROVIDERS_PRIORITY need to
	 * be higher than any other JVM hook that may require security providers.
	 */
	static final int RESET_CRIUSEC_PRIORITY = CONCURRENT_MODE_HOOK_PRIORITY_HIGHEST + 100;
	static final int RESTORE_SECURITY_PROVIDERS_PRIORITY = CONCURRENT_MODE_HOOK_PRIORITY_HIGHEST + 100;
```
Concurrent mode priority is from `0` to `99`.
Single thread mode priority is from `100` for `USER_HOOKS_PRIORITY`, `199` for `RESET_CRIUSEC_PRIORITY` and `RESTORE_SECURITY_PROVIDERS_PRIORITY` which are higher than others.
`UnsupportedOperationException` is to be thrown when the priority specified in `CONCURRENT_MODE` is not in the range of `CONCURRENT_MODE_HOOK_PRIORITY_LOWEST` and `CONCURRENT_MODE_HOOK_PRIORITY_HIGHEST`.

Reply:
I dont think we need different priorities for concurrent and single thread, the documentation explicitly says that single thread is higher priority